sw/device/lib/base/memory_perftest.c

```diff
@@ -0,0 +1,308 @@
+  static uint32_t seed = 0xbeefbeef;
+  seed *= 13;
+  for (size_t i = 0; i < len; ++i) {
+    buf[i] = i ^ 117 ^ (seed & 0xff);
+  }
+}
+
+static inline void fill_buf_zeroes(uint8_t *buf, size_t len) {
+  memset(buf, 0, len);
+}
+
+__attribute__((noinline)) void test_memcpy(test_context_t *context) {
+  // Copy every prefix of `buf2` to `buf1`.
+  for (size_t i = 0; i < context->len; ++i) {
+    memcpy(context->buf1, context->buf2, i);
+  }
+}
+
+__attribute__((noinline)) void test_memset(test_context_t *context) {
+  for (size_t i = 0; i < context->len; ++i) {
+    memset(context->buf1, context->buf2[0], i);
+  }
+}
+
+__attribute__((noinline)) void test_memcmp(test_context_t *context) {
+  for (size_t i = 0; i < context->len; ++i) {
+    memcmp(context->buf1, context->buf2, i);
+  }
+}
+
+__attribute__((noinline)) void test_memrcmp(test_context_t *context) {
+  for (size_t i = 0; i < context->len; ++i) {
+    memrcmp(context->buf1, context->buf2, i);
+  }
+}
+
+__attribute__((noinline)) void test_memchr(test_context_t *context) {
+  for (size_t i = 0; i < context->len; ++i) {
+    memchr(context->buf1, context->buf2[i], context->len);
+  }
+}
+
+__attribute__((noinline)) void test_memrchr(test_context_t *context) {
+  for (size_t i = 0; i < context->len; ++i) {
+    memrchr(context->buf1, context->buf2[i], context->len);
+  }
+}
+
+static size_t strlen(const char *string) {
+  const void *terminator = memchr(string, 0, SIZE_MAX);
+  ptrdiff_t diff = (uintptr_t)terminator - (uintptr_t)string;
+  CHECK(diff > 0);
+  return (size_t)diff;
+}
+
+static void print_padded(const char *string, size_t pad_to) {
+  base_printf("%s", string);
+  size_t label_len = strlen(string);
+  CHECK(pad_to >= label_len);
+  for (size_t padding = 0; padding < pad_to - label_len; padding++) {
+    base_printf(" ");
+  }
+}
+
+OTTF_DEFINE_TEST_CONFIG();
+
+bool test_main(void) {
+  static const perf_test_t kPerfTests[] = {
+      {
+          .label = "memcpy",
+          .setup_buf1 = &fill_buf_arbitrary_values,
+          .setup_buf2 = &fill_buf_arbitrary_values,
+          .func = &test_memcpy,
+          .num_reps = 4,
+      },
+      {
+          .label = "memcpy_zeroes",
+          .setup_buf1 = &fill_buf_arbitrary_values,
+          .setup_buf2 = &fill_buf_zeroes,
+          .func = &test_memcpy,
+          .num_reps = 4,
+      },
+      {
+          .label = "memset",
+          .setup_buf1 = &fill_buf_zeroes,
+          .setup_buf2 = &fill_buf_arbitrary_values,
+          .func = &test_memset,
+          .num_reps = 4,
+      },
+      {
+          .label = "memcmp",
+          .setup_buf1 = &fill_buf_arbitrary_values,
+          .setup_buf2 = &fill_buf_arbitrary_values,
+          .func = &test_memcmp,
+          .num_reps = 4,
+      },
+      {
+          .label = "memrcmp",
+          .setup_buf1 = &fill_buf_arbitrary_values,
+          .setup_buf2 = &fill_buf_arbitrary_values,
+          .func = &test_memrcmp,
+          .num_reps = 4,
+      },
```

Comment:
Given the setup for the `memcmp` family of functions, isn't memcmp likely to exit after comparing only a byte or two?  To test memcmp on long strides you're going to need buffers that are filled with the same sequence.

Reply:
Thanks for pointing that out, I haven't really paid attention to the quality of the inputs after memset. For memcmp-like functions, I think it makes sense to copy the buffers, but make a single mutation in various locations.

Reply:
I'm going to run memcmp and memrcmp on all zeroes for now, but I'm leaving a TODO that I'd like to resolve before we merge this.